File: frontend/lib/api/itineraries.ts
import { api } from './client';

export type ItineraryData = {
  itinerary_id: string;
  clerk_user_id: string;
  name: string;
  city: string;
  stop_count: number;
  created_at: string;
};

export async function saveItinerary(args: {
  itineraryId: string;
  clerkUserId: string;
  name: string;
  city: string;
  stopCount: number;
}): Promise<{ itinerary: ItineraryData }> {
  return api.post('/itineraries', {
    itineraryId: args.itineraryId,
    clerkUserId: args.clerkUserId,
    name: args.name,
    city: args.city,
    stopCount: args.stopCount,
  });
}

export async function getItineraries(
  clerkUserId: string
): Promise<{ itineraries: ItineraryData[] }> {
  return api.get('/itineraries', { clerkUserId });
}

File: backend/itineraries.py
from __future__ import annotations
import json
import os
import uuid
from dataclasses import dataclass, field, asdict
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional


@dataclass
class Itinerary:
    itinerary_id: str
    clerk_user_id: str
    name: str
    city: str
    stop_count: int
    created_at: str = field(default_factory=lambda: datetime.now(timezone.utc).isoformat())


class ItineraryStore:
    """
    File-backed itinerary store, keyed by itinerary_id.
    """
    itineraries: Dict[str, Dict[str, Any]] = {}
    _path: Optional[str] = None

    @classmethod
    def init(cls, path: str) -> None:
        cls._path = path
        if os.path.exists(path):
            with open(path, "r") as f:
                cls.itineraries = json.load(f)
        else:
            cls.itineraries = {}

    @classmethod
    def save(cls, itinerary: Itinerary) -> None:
        cls.itineraries[itinerary.itinerary_id] = asdict(itinerary)
        cls._write()

    @classmethod
    def get(cls, itinerary_id: str) -> Optional[Dict[str, Any]]:
        return cls.itineraries.get(itinerary_id)

    @classmethod
    def get_by_user(cls, clerk_user_id: str) -> List[Dict[str, Any]]:
        return [
            i for i in cls.itineraries.values()
            if i["clerk_user_id"] == clerk_user_id
        ]

    @classmethod
    def _write(cls) -> None:
        if cls._path:
            os.makedirs(os.path.dirname(cls._path), exist_ok=True)
            with open(cls._path, "w") as f:
                json.dump(cls.itineraries, f, indent=2)
